Code/GraphMol/Aromaticity.cpp

```diff
@@ -619,30 +648,107 @@ int countAtomElec(const Atom *at) {
 }
 
 namespace {
-// use minRingSize=0 or maxRingSize=0 to ignore these constraints
-int aromaticityHelper(RWMol &mol, unsigned int minRingSize,
-                      unsigned int maxRingSize, bool includeFused) {
-  // FIX: we will assume for now that if the input molecule came
-  // with aromaticity information it is correct and we will not
-  // touch it. Loop through the atoms and check if any atom has
-  // arom stuff set.  We may want check this more carefully later
-  // and start from scratch if necessary
-  ROMol::AtomIterator ai;
-  for (ai = mol.beginAtoms(); ai != mol.endAtoms(); ai++) {
-    if ((*ai)->getIsAromatic()) {
-      // found aromatic info
-      return -1;
+int mdlAromaticityHelper(RWMol &mol, const VECT_INT_VECT &srings) {
+  int narom = 0;
+  // loop over all the atoms in the rings that can be candidates
+  // for aromaticity
+  // Atoms are candidates if
+  //   - it is part of ring
+  //   - has one or more electron to donate or has empty p-orbitals
+  int natoms = mol.getNumAtoms();
+  boost::dynamic_bitset<> acands(natoms);
+  boost::dynamic_bitset<> aseen(natoms);
+  VECT_EDON_TYPE edon(natoms);
+
+  VECT_INT_VECT cRings;  // holder for rings that are candidates for aromaticity
+  for (auto &sring : srings) {
+    bool allAromatic = true;
+    bool allDummy = true;
+    for (auto firstIdx : sring) {
+      Atom *at = mol.getAtomWithIdx(firstIdx);
+
+      if (allDummy && at->getAtomicNum() != 0) {
+        allDummy = false;
+      }
+
+      if (aseen[firstIdx]) {
+        if (!acands[firstIdx]) allAromatic = false;
+        continue;
+      }
+      aseen[firstIdx] = 1;
+
+      // now that the atom is part of ring check if it can donate
+      // electron or has empty orbitals. Record the donor type
+      // information in 'edon' - we will need it when we get to
+      // the Huckel rule later
+      edon[firstIdx] = getAtomDonorTypeArom(at, false);
+      // we only accept one electron donors?
+      if (edon[firstIdx] != OneElectronDonorType) {
+        allAromatic = false;
+        continue;
+      }
+      bool allowThirdRow = false;
```

Comment:
These should be const, and I appreciate the effort!

Reply:
you won me over on this one... it makes the code much easier to read
No less error prone (since you can always provide them out of order) but certainly easier to figure out what is intended.